File: src/lib/combinatorialAnswerGrammar.ts
/**
 * Combinatorial disambiguation answer grammars: one panel per atomic value,
 * independent matching, engine resolves catalog value-set keys from matched atoms.
 */
import type { GrammarEntry } from './analysisTypes';
import type { DisambiguationQuestionStyle } from './disambiguationPlanTypes';
import {
  compileGrammarRegex,
  groupNameFromSlotSegment,
  normalizeGrammarEntry,
  validateGrammarRegex,
} from './grammarNormalize';
import {
  escapeRegexLiteral,
  expandSegmentVariants,
  normalizeSynonymList,
  sortSynonymsAlphabetically,
  type GrammarEditorPanel,
} from './grammarSynonyms';
import { isNoneOption } from './turnAnswerGrammar';
import {
  normalizeValueList,
  parseValueSetKey,
  VALUE_SET_SEPARATOR,
  valueSetContainsAll,
  valueSetKey,
} from './valueSet';

/** Extracts the pattern body of a named capture group from a regex string. */
export function extractNamedGroupPattern(regex: string, groupName: string): string | null {
  const marker = `(?<${groupName}>`;
  const idx = regex.indexOf(marker);
  if (idx < 0) return null;

  let depth = 1;
  let i = idx + marker.length;
  const start = i;
  while (i < regex.length && depth > 0) {
    const ch = regex[i]!;
    if (ch === '(' && regex[i - 1] !== '\\') depth += 1;
    else if (ch === ')' && regex[i - 1] !== '\\') depth -= 1;
    i += 1;
  }
  if (depth !== 0) return null;
  return regex.slice(start, i - 1);
}

/** True when disambiguation options need atomic combinatorial answer grammar. */
export function shouldUseCombinatorialAnswerGrammar(
  options: readonly string[],
  style?: DisambiguationQuestionStyle,
): boolean {
  if (style === 'optional_include' || style === 'ask_age') return false;

  const cleaned = options
    .map((o) => o.trim())
    .filter(Boolean)
    .filter((o) => !isNoneOption(o));
  if (cleaned.length === 0) return false;

  const hasCombined = cleaned.some((o) => o.includes(VALUE_SET_SEPARATOR));
  if (!hasCombined) return false;

  return extractAtomicTokensFromOptions(cleaned).length >= 2;
}

/** Distinct atomic tokens from all option value-set keys (longest first). */
export function extractAtomicTokensFromOptions(options: readonly string[]): string[] {
  const seen = new Set<string>();
  const atoms: string[] = [];

  for (const optionKey of options) {
    if (isNoneOption(optionKey)) continue;
    for (const token of parseValueSetKey(optionKey)) {
      const key = token.toLowerCase();
      if (seen.has(key)) continue;
      seen.add(key);
      atoms.push(token);
    }
  }

  return atoms.sort((a, b) => {
    const byLen = b.length - a.length;
    if (byLen !== 0) return byLen;
    return a.localeCompare(b, 'it');
  });
}

function atomGroupName(atom: string, usedNames: Set<string>): string {
  let base = groupNameFromSlotSegment(atom) || 'atom';
  let groupName = base;
  let suffix = 0;
  while (usedNames.has(groupName)) {
    groupName = `${base}_${suffix}`;
    suffix += 1;
  }
  usedNames.add(groupName);
  return groupName;
}

function defaultSynonymsForAtom(atom: string): string[] {
  return sortSynonymsAlphabetically(
    normalizeSynonymList([atom, ...expandSegmentVariants(atom)]),
  );
}

/** Builds synonym panels for atomic tokens (combinatorial mode). */
export function buildCombinatorialAnswerGrammarPanels(
  options: readonly string[],
  grammar: GrammarEntry | null | undefined,
): GrammarEditorPanel[] {
  const atoms = extractAtomicTokensFromOptions(options);
  const entry = grammar?.combinatorial ? normalizeGrammarEntry(grammar) : null;

  return atoms.map((atom) => {
    let synonyms = defaultSynonymsForAtom(atom);
    if (entry?.regex?.trim()) {
      const groupName = Object.entries(entry.mappings).find(([, v]) => v === atom)?.[0];
      if (groupName) {
        const pattern = extractNamedGroupPattern(entry.regex, groupName);
        if (pattern) {
          const fromPattern = pattern
            .split('|')
            .map((s) => s.replace(/\\(.)/g, '$1').trim())
            .filter(Boolean);
          if (fromPattern.length > 0) {
            synonyms = sortSynonymsAlphabetically(normalizeSynonymList(fromPattern));
          }
        }
      }
    }

    return {
      targetPath: atom,
      label: atom,
      isParent: false,
      synonyms,
    };
  });
}

/** Compiles a combinatorial answer grammar from atomic synonym panels. */
export function compileCombinatorialAnswerGrammarFromPanels(
  panels: GrammarEditorPanel[],
): GrammarEntry {
  const parts: string[] = [];
  const mappings: Record<string, string> = {};
  const usedNames = new Set<string>();

  for (const panel of panels) {
    const cleaned = normalizeSynonymList(panel.synonyms);
    if (cleaned.length === 0) continue;

    const groupName = atomGroupName(panel.targetPath, usedNames);
    parts.push(`(?<${groupName}>${cleaned.map(escapeRegexLiteral).join('|')})`);
    mappings[groupName] = panel.targetPath;
  }

  if (parts.length === 0) {
    throw new Error('Inserisci almeno un sinonimo in un pannello atomico');
  }

  const entry = normalizeGrammarEntry({
    regex: parts.join('|'),
    mappings,
    combinatorial: true,
  });
  const validation = validateGrammarRegex(entry.regex, entry.mappings);
  if (!validation.valid) {
    throw new Error(validation.error ?? 'Grammatica combinatoria non valida');
  }
  return entry;
}

/** Auto-generates combinatorial answer grammar from catalog option keys. */
export function compileCombinatorialAnswerGrammar(options: readonly string[]): GrammarEntry | null {
  if (!shouldUseCombinatorialAnswerGrammar(options)) return null;
  try {
    const panels = buildCombinatorialAnswerGrammarPanels(options, null);
    return compileCombinatorialAnswerGrammarFromPanels(panels);
  } catch {
    return null;
  }
}

/** Matches every atomic value mentioned independently in the utterance. */
export function matchAllCombinatorialAtoms(text: string, grammar: GrammarEntry): string[] {
  const trimmed = text.trim().toLowerCase();
  if (!trimmed || !grammar.regex?.trim()) return [];

  const entry = normalizeGrammarEntry(grammar);
  const matched: string[] = [];

  for (const [groupName, atom] of Object.entries(entry.mappings)) {
    const pattern = extractNamedGroupPattern(entry.regex, groupName);
    if (!pattern) continue;
    try {
      const re = compileGrammarRegex(`(?<${groupName}>${pattern})`);
      if (!re.test(trimmed)) continue;
      if (!matched.some((m) => m.toLowerCase() === atom.toLowerCase())) {
        matched.push(atom);
      }
    } catch {
      continue;
    }
  }

  return normalizeValueList(matched);
}

/**
 * Resolves a catalog option key from mentioned atomic values.
 * Exact match first, then maximal compatible combination (product rule).
 */
export function resolveOptionKeyFromMatchedAtoms(
  mentioned: readonly string[],
  options: readonly string[],
): string | null {
  const normalized = normalizeValueList(mentioned);
  if (normalized.length === 0) return null;

  const cleaned = options.map((o) => o.trim()).filter(Boolean);
  const mentionedKey = valueSetKey(normalized);
  const exact = cleaned.find((o) => o.toLowerCase() === mentionedKey.toLowerCase());
  if (exact) return exact;

  let bestKey: string | null = null;
  let bestCount = -1;

  for (const optionKey of cleaned) {
    if (isNoneOption(optionKey)) continue;
    const optionValues = parseValueSetKey(optionKey);
    if (!valueSetContainsAll(optionValues, normalized)) continue;
    if (optionValues.length > bestCount) {
      bestCount = optionValues.length;
      bestKey = optionKey;
    }
  }

  return bestKey;
}

export function isCombinatorialAnswerGrammar(
  grammar: GrammarEntry | null | undefined,
): boolean {
  return grammar?.combinatorial === true;
}

File: dialog-engine/src/DialogEngine/CombinatorialAnswerGrammar.vb
''' <summary>
''' Combinatorial disambiguation answer grammars: independent atomic matching,
''' value-set key resolution delegated to DisambiguationTokenMatch.
''' </summary>
Imports System.Text.RegularExpressions

Public Module CombinatorialAnswerGrammar

    Public Function IsCombinatorialGrammar(grammar As Models.CategoryGrammar) As Boolean
        Return grammar IsNot Nothing AndAlso If(grammar.Combinatorial, False)
    End Function

    ''' <summary>Returns every atomic mapping value independently matched in the utterance.</summary>
    Public Function MatchAllAtoms(text As String, grammar As Models.CategoryGrammar) As List(Of String)
        Dim mentioned As New List(Of String)()
        If grammar Is Nothing OrElse String.IsNullOrWhiteSpace(grammar.Regex) Then Return mentioned
        If grammar.Mappings Is Nothing OrElse grammar.Mappings.Count = 0 Then Return mentioned

        Dim trimmed = If(text, String.Empty).Trim().ToLowerInvariant()
        If trimmed.Length = 0 Then Return mentioned

        For Each groupName In grammar.Mappings.Keys
            Dim atom = grammar.Mappings(groupName)
            If String.IsNullOrWhiteSpace(atom) Then Continue For

            Dim pattern = ExtractNamedGroupPattern(grammar.Regex, groupName)
            If String.IsNullOrWhiteSpace(pattern) Then Continue For

            Try
                Dim re As New Regex(
                    $"(?<!\w)(?:{pattern})(?!\w)",
                    RegexOptions.IgnoreCase Or RegexOptions.CultureInvariant)
                If Not re.IsMatch(trimmed) Then Continue For
                If Not mentioned.Any(Function(m) String.Equals(m, atom, StringComparison.OrdinalIgnoreCase)) Then
                    mentioned.Add(atom.Trim())
                End If
            Catch
                Continue For
            End Try
        Next

        Return ValueSetOps.NormalizeAttributoValues(mentioned)
    End Function

    ''' <summary>Matches atoms via grammar and resolves the catalog option key.</summary>
    Public Function MatchAndResolveOptionKey(
        text As String,
        grammar As Models.CategoryGrammar,
        options As IList(Of String)
    ) As String
        Dim mentioned = MatchAllAtoms(text, grammar)
        If mentioned.Count = 0 Then Return Nothing
        Return DisambiguationTokenMatch.ResolveOptionKeyFromMentionedAtoms(mentioned, options)
    End Function

    Friend Function ExtractNamedGroupPattern(regex As String, groupName As String) As String
        If String.IsNullOrWhiteSpace(regex) OrElse String.IsNullOrWhiteSpace(groupName) Then Return Nothing

        Dim marker = "(?<" & groupName & ">"
        Dim idx = regex.IndexOf(marker, StringComparison.Ordinal)
        If idx < 0 Then Return Nothing

        Dim depth = 1
        Dim i = idx + marker.Length
        Dim start = i
        While i < regex.Length AndAlso depth > 0
            Dim ch = regex(i)
            If ch = "("c AndAlso (i = 0 OrElse regex(i - 1) <> "\"c) Then
                depth += 1
            ElseIf ch = ")"c AndAlso (i = 0 OrElse regex(i - 1) <> "\"c) Then
                depth -= 1
            End If
            i += 1
        End While

        If depth <> 0 Then Return Nothing
        Return regex.Substring(start, i - start - 1)
    End Function

End Module
